File: backend/src/main/java/com/upi/gateway/backend/service/PaymentService.java
package com.upi.gateway.backend.service;

import com.upi.gateway.backend.dto.PaymentInitiateRequest;
import com.upi.gateway.backend.dto.PaymentInitiateResponse;
import com.upi.gateway.backend.dto.TransactionStatusResponse;
import com.upi.gateway.backend.factory.PaymentProcessorFactory;
import com.upi.gateway.backend.model.Transaction;
import com.upi.gateway.backend.observer.TransactionObserver;
import com.upi.gateway.backend.observer.TransactionSubject;
import com.upi.gateway.backend.repository.TransactionRepository;
import com.upi.gateway.backend.strategy.PaymentProcessor;
import com.upi.gateway.backend.strategy.PaymentResult;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

@Service
@RequiredArgsConstructor
@Slf4j
public class PaymentService implements TransactionSubject {

    private final TransactionRepository transactionRepository;
    private final PaymentProcessorFactory processorFactory;
    private final List<TransactionObserver> observers = new ArrayList<>();

    @Override
    public void registerObserver(TransactionObserver observer) {
        observers.add(observer);
        log.info("Registered observer: {}", observer.getType());
    }

    @Override
    public void removeObserver(TransactionObserver observer) {
        observers.remove(observer);
        log.info("Removed observer: {}", observer.getType());
    }

    @Override
    public void notifyObservers(Transaction transaction) {
        log.info("Notifying {} observers for transaction {} with status {}",
                observers.size(), transaction.getId(), transaction.getStatus());

        for (TransactionObserver observer : observers) {
            try {
                observer.update(transaction);
            } catch (Exception e) {
                log.error("Error notifying observer {}: {}", observer.getType(), e.getMessage(), e);
            }
        }
    }

    @Transactional
    public PaymentInitiateResponse initiatePayment(PaymentInitiateRequest request) {
        log.info("Initiating payment for merchant: {} with method: {}",
                request.getMerchantId(), request.getPaymentMethod());

        Optional<PaymentProcessor> processorOpt = processorFactory.getProcessorByMethod(request.getPaymentMethod());
        if (processorOpt.isEmpty()) {
            throw new IllegalArgumentException("Unsupported payment method: " + request.getPaymentMethod());
        }

        Transaction transaction = Transaction.builder()
                .merchantId(request.getMerchantId())
                .amount(request.getAmount())
                .paymentMethod(request.getPaymentMethod().toUpperCase())
                .status(Transaction.TransactionStatus.CREATED)
                .merchantTransactionId(request.getMerchantTransactionId())
                .customerEmail(request.getCustomerEmail())
                .customerPhone(request.getCustomerPhone())
                .callbackUrl(request.getCallbackUrl())
                .build();

        transaction = transactionRepository.save(transaction);
        log.info("Created transaction with ID: {}", transaction.getId());

        notifyObservers(transaction);

        transaction.setStatus(Transaction.TransactionStatus.INITIATED);
        transaction = transactionRepository.save(transaction);

        notifyObservers(transaction);
        return PaymentInitiateResponse.from(transaction, "Payment initiated successfully");
    }

    @Transactional
    public void updateTransactionResult(UUID transactionId, PaymentResult result) {
        Optional<Transaction> transactionOpt = transactionRepository.findById(transactionId);
        if (transactionOpt.isPresent()) {
            Transaction transaction = transactionOpt.get();
            transaction.setStatus(result.getStatus());

            if (result.isSuccess() && result.getTransactionId() != null) {
                transaction.setUpiTransactionId(result.getTransactionId());
            }

            if (!result.isSuccess() && result.getFailureReason() != null) {
                transaction.setFailureReason(result.getFailureReason());
            }

            transaction = transactionRepository.save(transaction);
            log.info("Updated transaction {} with status: {}", transactionId, result.getStatus());

            notifyObservers(transaction);
        }
    }

    public Optional<TransactionStatusResponse> getTransactionStatus(UUID transactionId) {
        return transactionRepository.findById(transactionId)
                .map(TransactionStatusResponse::from);
    }

    public List<TransactionStatusResponse> getMerchantTransactions(Long merchantId) {
        return transactionRepository.findByMerchantIdOrderByCreatedAtDesc(merchantId)
                .stream()
                .map(TransactionStatusResponse::from)
                .toList();
    }

    public List<String> getAvailablePaymentMethods() {
        return processorFactory.getAvailablePaymentMethods();
    }
}

File: payment-service/services/PaymentService.js
import { createClient } from '@supabase/supabase-js';
import { PaymentProcessorFactory } from '../factory/PaymentProcessorFactory.js';
import { v4 as uuidv4 } from 'uuid';

export class PaymentService {
    constructor() {
        this.supabase = createClient(
            process.env.SUPABASE_URL,
            process.env.SUPABASE_ANON_KEY
        );
        this.processorFactory = new PaymentProcessorFactory();
    }

    async initiatePayment(request) {
        console.log(`Initiating payment for merchant: ${request.merchantId} with method: ${request.paymentMethod}`);
        
        // Validate payment method
        const processor = this.processorFactory.getProcessorByMethod(request.paymentMethod);
        if (!processor) {
            throw new Error(`Unsupported payment method: ${request.paymentMethod}`);
        }

        // Create transaction with CREATED status
        const transaction = {
            id: uuidv4(),
            merchant_id: request.merchantId,
            amount: request.amount,
            payment_method: request.paymentMethod.toUpperCase(),
            status: 'CREATED',
            merchant_transaction_id: request.merchantTransactionId,
            customer_email: request.customerEmail,
            customer_phone: request.customerPhone
        };

        const { data: createdTransaction, error: createError } = await this.supabase
            .from('transactions')
            .insert([transaction])
            .select()
            .single();

        if (createError) {
            console.error('Error creating transaction:', createError);
            throw new Error('Failed to create transaction');
        }

        console.log(`Created transaction with ID: ${createdTransaction.id}`);

        // Update status to INITIATED
        const { error: updateError } = await this.supabase
            .from('transactions')
            .update({ status: 'INITIATED' })
            .eq('id', createdTransaction.id);

        if (updateError) {
            console.error('Error updating transaction status:', updateError);
        }

        // Process payment asynchronously
        this.processPaymentAsync(createdTransaction, processor);

        return {
            transactionId: createdTransaction.id,
            merchantId: createdTransaction.merchant_id,
            amount: createdTransaction.amount,
            paymentMethod: createdTransaction.payment_method,
            status: 'INITIATED',
            message: 'Payment initiated successfully',
            createdAt: createdTransaction.created_at,
            merchantTransactionId: createdTransaction.merchant_transaction_id
        };
    }

    async processPaymentAsync(transaction, processor) {
        try {
            // Update status to PROCESSING
            await this.supabase
                .from('transactions')
                .update({ status: 'PROCESSING' })
                .eq('id', transaction.id);

            // Process payment
            const result = await processor.process(transaction);
            await this.updateTransactionResult(transaction.id, result);
        } catch (error) {
            console.error(`Payment processing failed for transaction: ${transaction.id}`, error);
            const failureResult = {
                success: false,
                status: 'FAILED',
                failureReason: `Processing error: ${error.message}`
            };
            await this.updateTransactionResult(transaction.id, failureResult);
        }
    }

    async updateTransactionResult(transactionId, result) {
        const updateData = {
            status: result.status
        };

        if (result.success && result.transactionId) {
            updateData.upi_transaction_id = result.transactionId;
        }

        if (!result.success && result.failureReason) {
            updateData.failure_reason = result.failureReason;
        }

        const { error } = await this.supabase
            .from('transactions')
            .update(updateData)
            .eq('id', transactionId);

        if (error) {
            console.error('Error updating transaction result:', error);
        } else {
            console.log(`Updated transaction ${transactionId} with status: ${result.status}`);
        }
    }

    async getTransactionStatus(transactionId) {
        const { data, error } = await this.supabase
            .from('transactions')
            .select('*')
            .eq('id', transactionId)
            .single();

        if (error) {
            console.error('Error fetching transaction:', error);
            return null;
        }

        return data;
    }

    async getMerchantTransactions(merchantId) {
        const { data, error } = await this.supabase
            .from('transactions')
            .select('*')
            .eq('merchant_id', merchantId)
            .order('created_at', { ascending: false });

        if (error) {
            console.error('Error fetching merchant transactions:', error);
            return [];
        }

        return data;
    }

    getAvailablePaymentMethods() {
        return this.processorFactory.getAvailablePaymentMethods();
    }
}